### ISSUE TITLE
Why does the C libmpi.so have a gfortran/fortran dependence?

### ISSUE THREAD

**Issue by harisubramoni** [pmodels/mpich#4130]
Hi, All.

I built MPICH 3.3 in two modes 1) Default 2) After adding --disable-fortran at configure time.

Default
----------
Build flags: CC=gcc CXX=g++ FC=gfortran F77=gfortran CFLAGS=-pipe

[subramon@head git]$ ldd install-dir/lib/libmpi.so | grep fort
        libgfortran.so.5 => /opt/gcc/8.3.0/lib64/libgfortran.so.5 (0x00002b46fdd4a000)

Without Fortran
-------------------- 
Build flags CC=gcc CXX=g++ FC=gfortran F77=gfortran CFLAGS=-pipe --disable-fortran

[subramon@head mv2-git]$ ldd install-dir-nofortran/lib/libmpi.so | grep fort
<Nothing>

This leads me to believe that during the link step, the build system is linking in the Fortran libraries into libmpi.so. Could you please let me know why this is happening? Is this an intentional feature? If so, why is it there given that there is a separate libmpifort.so?

```
[subramon@head git]$ git branch
* (detached from v3.3)
  master
[subramon@head git]$ git log
commit 9db5d2e80c0e88e33e82bf08b849f1dd3a2d919b
Author: Ken Raffenetti <raffenet@mcs.anl.gov>
Date:   Wed Nov 21 11:20:29 2018 -0600

    release: Update shortlog link in CHANGES file

    No reviewer.
```

Thx,
Hari.

**Comment by hzhou** [comment]
When FORTRAN binding is enabled, `libmpi.so` is linked with FORTRAN compiler, e.g. `gfortran`. to ensure FORTRAN dependency is being linked in. I am not sure this is necessary today -- maybe I should investigate -- but at time we must have found this to be necessary to make the fortran binding work.

@harisubramoni Why this is causing an issue for you?

**Comment by harisubramoni** [comment]
Hi, Hui.

Thanks for taking a look at this. This was a question raised by one of our users.

The only question is why is linking fortran dependencies into libmpi.so required when there is a separate libmpifort.so available when fortran is not disabled? Can't users requiring fortran use this instead of libmpi.so? Could you please let me know if there some other issue that I am missing here?

Thx,
Hari.

**Comment by hzhou** [comment]
The decision is made in `Makefile.am` (https://github.com/pmodels/mpich/blob/d5504b874ddfb9e95c059be6eb62b39a1b930d81/Makefile.am#L69). We certainly can discard the `if BUILD_F77_BINDING` case and always use `$(CCLD)` for `lib_lib@MPILIBNAME@_la_LINK`. Then `libmpi.so` will not require FORTRAN linkage. As far as I know, there is only one fortran code -- `src/bindings/fortran/mpif_h/setbotf.f` is in `libmpi.so` which only provides some global variables. And I believe it works fine for typicl tool chain, such as linux/elf/gnu.

**Comment by harisubramoni** [comment]
Hi, Hui.

Thanks for the detailed feedback. If it would be possible and okay, can this be done?

Best,
Hari.

**Comment by harisubramoni** [comment]
Hi, Hui.

I am just following up here to see if there is any update here. We are trying to finalize a release soon and was wondering if there is a self contained patch we can take in.

Thx,
Hari.

**Comment by hzhou** [comment]
Hi Hari,

Both GitHub PR #4140 and PR #4149 seems work and clears the fortran dependence. #4140 is trivia in code change, simply removes the dependency from linker, which is most likely what user is seeking. However, it contains object from fortran code, so there is a risk that it breaks the toolchain and officially we are reluctant to support. #4149 went deeper refactoring and separates the code more thoroughly. As such, we are having difficulty to get it thoroughly reviewed. In short, we have trouble to make the cut (any external reviews may help). As far as personally, I think either PR is a good candidate for a release to incorporate.

Sorry that we couldn’t make it easy :)
—
Hui Zhou



Links:
https://github.com/pmodels/mpich/pull/4140
https://github.com/pmodels/mpich/pull/4149





On Dec 9, 2019, at 1:01 PM, Hari Subramoni <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi, Hui.

I am just following up here to see if there is any update here. We are trying to finalize a release soon and was wondering if there is a self contained patch we can take in.

Thx,
Hari.

**Comment by hzhou** [comment]
Very related: https://bitbucket.org/mpi4py/mpi4py/issues/162/mpi4py-initialization-breaks-fortran

**Comment by hzhou** [comment]
fixed by #4149
